Source/WebCore/dom/ElementTextDirection.cpp

```diff
@@ -0,0 +1,326 @@
+{
+    if (!textFormControlElement.dirAutoUsesValue())
+        return computeElementContainedTextAutoDirection(textFormControlElement);
+
+    // Specs: The directionality of the auto-directionality form-associated
+    // element is calculated from its value() text.
+    // Specs: If element's value is not the empty string, then return 'ltr'.
+    if (auto value = textFormControlElement.value(); !value.isEmpty())
+        return computeTextDirectionFromText(value).value_or(TextDirection::LTR);
+    return std::nullopt;
+}
+
+std::optional<TextDirection> computeElementAutoTextDirection(const Element& element)
+{
+    if (auto* textFormControl = dynamicDowncast<HTMLTextFormControlElement>(element))
+        return computeTextFormControlElementTextDirection(*textFormControl);
+
+    if (auto* slotElement = dynamicDowncast<HTMLSlotElement>(element))
+        return computeSlotElementTextDirection(*slotElement);
+
+    return computeElementContainedTextAutoDirection(element);
+}
+
+static std::optional<TextDirection> computeElementTextDirection(const Element& element, TextDirectionality textDirectionality)
+{
+    // Specs: https://html.spec.whatwg.org/multipage/dom.html#attr-dir
+    switch (textDirectionality) {
+    case TextDirectionality::LTR:
+        return TextDirection::LTR;
+    case TextDirectionality::RTL:
+        return TextDirection::RTL;
+
+    case TextDirectionality::Auto:
+        // Specs: Return the auto directionality of element.
+        // Specs: If directionality is null, then return 'ltr'.
+        return computeElementAutoTextDirection(element).value_or(TextDirection::LTR);
+
+    case TextDirectionality::Undefined:
+        // Specs: If element is a bdi element, then return the auto directionality of element.
+        // Specs: If directionality is null, then return 'ltr'.
+        if (is<HTMLBDIElement>(element))
+            return computeElementAutoTextDirection(element).value_or(TextDirection::LTR);
+
+        // Specs: If element is an input element whose type attribute is in the Telephone state
+        // then return LTR;
+        if (auto* input = dynamicDowncast<HTMLInputElement>(element); input && input->isTelephoneField())
+            return TextDirection::LTR;
+
+        // Specs: If the parent is a shadow root, then return the directionality of parent's host.
+        // Specs: If the parent is an element, then return the directionality of parent.
+        if (RefPtr parent = element.parentOrShadowHostElement())
+            return computeElementTextDirection(*parent);
+
+        return std::nullopt;
+    }
+
+    return std::nullopt;
+}
+
+std::optional<TextDirection> computeElementTextDirection(const Element& element)
+{
+    return computeElementTextDirection(element, elementTextDirectionality(element));
+}
+
+static void updateElementSelfTextDirection(Element& element, std::optional<TextDirection> direction, Element* exclude = nullptr)
```

Comment:
This function name is confusing. It says self but then it updates the descendants of its shadow tree.

Reply:
I assumed the shadow tree is something owned and hidden by the node since it is not exposed directly by DOM.

How about `updateElementAndShadowTextDirection()`?